### DISCUSSION TITLE
Which client software to use for windows?

### DISCUSSION THREAD

**Discussion by skuskusku** [dorssel/usbipd-win#908] [Q&A]
Hello Frans,

it is unclear to me if usbipd-win is suitable for my use case: I want to remote a digicert code signing usb token to a windows build vm. I don't care if the server side is running on windows or linux, but I need client side access from a windows 10 build vm to the server running usbipd-win or similar. 

What I haven't understood is: Does the downloadable MSI of yours only supply the server side of USB-remoting or does it contain the necessary software for clients as well? If not, can you recommend a client package for windows? From my investigations so far, all IP/USB windows based drivers are in not so a mature state, lots haven't been updated for ages and some lack properly signed drivers. Can you share your experience? Stability is of uttermost importance for my build VMs.

Any help appreciated,

-- 
Stefan

**Comment by dorssel** [comment]
You are correct. `usbipd-win` is only the server side of the protocol. Client tooling for Windows appears to exist; some users reported success. But I don't use Windows clients myself, so I wouldn't know. As reported in other issues, and in the README, client side is outside the scope of this project. Maybe search through the older (possibly closed) issues/discussions for user stories.

**Comment by Sucareto** [comment]
I also hope to find a suitable Windows client software. I tried [usbip-win](https://github.com/cezanne/usbip-win) and it worked fine.

I shared a COM device within the LAN using usbipd-win. With usbip-win connection, it can be used successfully without enabling testsigning.

Command Reference:

Server:
``` powershell
winget install usbipd
# Found device busid or vid:pid
usbipd list
# Share my ESP32 com
usbipd bind -i 1a86:7523
```

Client:
``` powershell
# Download and unzip usbip-win
# https://github.com/cezanne/usbip-win/releases/download/v0.3.6-dev/usbip-win-0.3.6-dev.zip
usbip.exe install
# List shared devices on the server
usbip.exe list -r 192.168.0.100
# Attach device
usbip.exe attach -r 192.168.0.100 -b 3-2
# Show imported USB devices
usbip.exe port
# Detach device
usbip.exe detach -p 00
```
Additionally, you can also take a look at [usbip-win2](https://github.com/vadimgrn/usbip-win2), which reportedly includes a GUI. However, I haven't tested it myself because my device cannot enable testsigning.
If there's a better method, please let me know. Thanks in advance.

**Reply by patrikhuber** [reply]
@Sucareto The original https://github.com/cezanne/usbip-win also requires test-signing according to their readme, how did you get it to work without enabling test-signing?

**Reply by Sucareto** [reply]
> @Sucareto根据他们的自述文件，原始的https://github.com/cezanne/usbip-win也需要测试签名，您如何在不启用测试签名的情况下让它工作？

When I wrote the above comment, I did not enable test signing, and I just installed cezanne/usbip-win and it worked.
Did you find any problems while using it?

**Comment by patrikhuber** [comment]
I'm wondering, on the client-side, is it possible to use `usbip` from WSL2 to connect to the remote USB device, and then somehow "share" or "forward" the device from inside WSL2 to Windows?

Any ideas? Perhaps @dorssel?

**Reply by dorssel** [reply]
Unfortunately, Linux `usbipd` does not allow "re-forwarding" a USB device. It detects that a device is imported via USBIP and then refuses to bind it. Probably to prevent "loops". That's a "feature" of Linux USBIP.
